go/vt/vtgate/planbuilder/querytree_transformers.go

```diff
@@ -309,8 +315,15 @@ func mergeUnionLogicalPlans(ctx *planningContext, left logicalPlan, right logica
 		return nil
 	}
 
+	// safe to append system table schema and system table names, since either the routing will match or either side would be throwing an error
+	// during run-time which we want to preserve. For example outer side has User in sys table schema and inner side has User and Main in sys table schema
+	// Inner might end up throwing an error at runtime, but if it doesn't then it is safe to merge.
 	if canMergeUnionPlans(ctx, lroute, rroute) {
 		lroute.Select = &sqlparser.Union{Left: lroute.Select, Distinct: false, Right: rroute.Select}
+		lroute.eroute.SysTableTableSchema = append(lroute.eroute.SysTableTableSchema, rroute.eroute.SysTableTableSchema...)
+		for k, v := range rroute.eroute.SysTableTableName {
+			lroute.eroute.SysTableTableName[k] = v
+		}
```

Comment:
nit: we can export this and lines 149 to 152 to a method

Reply:
Fixed